Dispatch GlobalObjectAvailable to subframes when adding user scripts in top frame only

laphets opened the pull request:
#### 4b458ff27f4460ee0e608f378b5e1af3b1949f41
<pre>
Dispatch GlobalObjectAvailable to subframes when adding user scripts in top frame only
<a href="https://bugs.webkit.org/show_bug.cgi?id=243606">https://bugs.webkit.org/show_bug.cgi?id=243606</a>
&lt;rdar://73500002&gt;

Reviewed by Alex Christensen, Timothy Hatcher and Chris Dumez.

Currently, after navigation completes, and add user script to top frame only,
dispatchGlobalObjectAvailable won&apos;t be called for its subframes. If later
runJavaScriptInFrameInScriptWorld is called directly in the subframe using the
same world as the user script added to top frame, dispatchGlobalObjectAvailable
won&apos;t be called for the subframe either. As a result, objects registered in
dispatchGlobalObjectAvailable callback won&apos;t be available for code that is
executed by runJavaScriptInFrameInScriptWorld in subframe.
The patch aims to address the case when a user script is added after navigation
completes, it fixes the issue by explicitly dispatching GlobalObjectAvailable
to all the subframes after adding a new content world.

* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addUserScriptInternal):
When adding a new content world, traverse all the subframes under mainframe,
call dispatchGlobalObjectAvailable on each of subframe using the given world.

Canonical link: <a href="https://commits.webkit.org/253337@main">https://commits.webkit.org/253337@main</a>
</pre>

xeenon reviewed (approved):
Looks fine to me. Would like the other reviewers to take a look still.

cdumez commented on Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
```diff
@@ -417,7 +418,12 @@ void WebUserContentController::addUserScriptInternal(InjectedBundleScriptWorld&
             auto& mainFrame = page.mainFrame();
             if (userScript.injectedFrames() == UserContentInjectedFrames::InjectInTopFrameOnly) {
-                mainFrame.injectUserScriptImmediately(world.coreWorld(), userScript);
+                for (auto* frame = &mainFrame; frame; frame = frame->tree().traverseNext(&mainFrame)) {
```
the `&mainFrame` parameter to `traverseNext()` is not needed since you are traversing the whole frame tree.

cdumez commented on Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
```diff
@@ -417,7 +418,12 @@ void WebUserContentController::addUserScriptInternal(InjectedBundleScriptWorld&
             auto& mainFrame = page.mainFrame();
             if (userScript.injectedFrames() == UserContentInjectedFrames::InjectInTopFrameOnly) {
-                mainFrame.injectUserScriptImmediately(world.coreWorld(), userScript);
+                for (auto* frame = &mainFrame; frame; frame = frame->tree().traverseNext()) {
```
It is a bit inefficient to check every frame to see if it's the main frame (when we know only the first one is). Would be more efficient to write like so:
```
mainFrame.injectUserScriptImmediately(world.coreWorld(), userScript);
for (auto* frame = mainFrame.tree().traverseNext(); frame; frame = frame->tree().traverseNext())
    frame->loader().client().dispatchGlobalObjectAvailable(world.coreWorld());
```

cdumez reviewed (approved):
LGTM with change.

achristensen07 commented on Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
```diff
@@ -418,6 +419,8 @@ void WebUserContentController::addUserScriptInternal(InjectedBundleScriptWorld&
             if (userScript.injectedFrames() == UserContentInjectedFrames::InjectInTopFrameOnly) {
                 mainFrame.injectUserScriptImmediately(world.coreWorld(), userScript);
+                for (auto* frame = mainFrame.tree().traverseNext(); frame; frame = frame->tree().traverseNext())
+                    frame->loader().client().dispatchGlobalObjectAvailable(world.coreWorld());
```
Yeah, this isn't the right place for this.

laphets committed:
Dispatch GlobalObjectAvailable to subframes when adding user scripts in top frame only
https://bugs.webkit.org/show_bug.cgi?id=243606
<rdar://73500002>

Reviewed by Alex Christensen, Timothy Hatcher and Chris Dumez.

Currently, after navigation completes, and add user script to top frame only,
dispatchGlobalObjectAvailable won't be called for its subframes. If later
runJavaScriptInFrameInScriptWorld is called directly in the subframe using the
same world as the user script added to top frame, dispatchGlobalObjectAvailable
won't be called for the subframe either. As a result, objects registered in
dispatchGlobalObjectAvailable callback won't be available for code that is
executed by runJavaScriptInFrameInScriptWorld in subframe.
The patch aims to address the case when a user script is added after navigation
completes, it fixes the issue by explicitly dispatching GlobalObjectAvailable
to all the subframes after adding a new content world.

* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addUserScriptInternal):
When adding a new content world, traverse all the subframes under mainframe,
call dispatchGlobalObjectAvailable on each of subframe using the given world.

Canonical link: https://commits.webkit.org/253337@main
